File: examples/pearltrees/demo_bookmark_filing.rs
// Demo: Bookmark Filing Assistant using semantic search and tree formatting
//
// This demonstrates Rust's bookmark filing capabilities:
// - Vector similarity search with ONNX embeddings
// - Tree context formatting (ancestors, siblings, children)
// - Semantic placement suggestions

mod importer;
mod crawler;
mod searcher;
mod embedding;

use searcher::PtSearcher;
use embedding::EmbeddingProvider;
use std::error::Error;

fn main() -> Result<(), Box<dyn Error>> {
    println!("=== Bookmark Filing Assistant Demo ===\n");
    println!("Loading database and embedding model...");

    let model_path = "models/all-MiniLM-L6-v2-safetensors/model.safetensors";
    let tokenizer_path = "models/all-MiniLM-L6-v2-safetensors/tokenizer.json";
    let db_path = "pt_ingest_test.redb";

    let embedder = EmbeddingProvider::new(model_path, tokenizer_path)?;
    let searcher = PtSearcher::new(db_path, embedder)?;

    println!("✓ Ready\n");

    // Test cases for bookmark filing
    let bookmarks = vec![
        "Article about quantum entanglement and its applications in quantum computing",
        "Tutorial on classical mechanics and Newton's laws of motion",
        "Research paper on electromagnetic waves and Maxwell's equations",
    ];

    for bookmark in &bookmarks {
        println!("{}", "=".repeat(100));
        println!();

        // Try with lower threshold to see results
        let result = searcher.find_bookmark_placements(bookmark, 3, 0.1)?;
        println!("{}", result);

        println!();
    }

    println!("{}", "=".repeat(100));
    println!("\n✓ Demo complete");

    Ok(())
}

File: examples/pearltrees/demo_bookmark_filing.cs
using System;
using LiteDB;
using UnifyWeaver.QueryRuntime;

// Demo: Bookmark Filing Assistant using semantic search and tree formatting
var dbPath = "pt_ingest_test.db";
var modelPath = "models/all-MiniLM-L6-v2/model.onnx";
var vocabPath = "models/all-MiniLM-L6-v2/vocab.txt";

Console.WriteLine("=== Bookmark Filing Assistant Demo ===\n");
Console.WriteLine("Loading database and embedding model...");

var embeddingProvider = new OnnxEmbeddingProvider(modelPath, vocabPath);
using var searcher = new PtSearcher(dbPath, embeddingProvider);

Console.WriteLine("✓ Ready\n");

// Test cases for bookmark filing
var bookmarks = new[]
{
    "Article about quantum entanglement and its applications in quantum computing",
    "Tutorial on classical mechanics and Newton's laws of motion",
    "Research paper on electromagnetic waves and Maxwell's equations"
};

foreach (var bookmark in bookmarks)
{
    Console.WriteLine(new string('=', 100));
    Console.WriteLine();

    var result = searcher.FindBookmarkPlacements(bookmark, topCandidates: 3, minScore: 0.35);
    Console.WriteLine(result);

    Console.WriteLine();
}

Console.WriteLine(new string('=', 100));
Console.WriteLine("\n✓ Demo complete");
